Can't build on macOS because `nix::unistd::daemon` doesn't exist

I don't think this is easily fixable, but I am building on macOS and get:

```
error[E0432]: unresolved import `nix::unistd::daemon`
```

I assume this is because `dameon` is not included in `nix::unistd` on macOS (as per [this line](https://github.com/nix-rust/nix/blob/3f637de8e027843e1b2c4e9a05f19b071ea00017/src/unistd.rs#L968)). I am not a Rust programmer; are there any quick fixes I could make on macOS to solve this?

Thanks

> 1. OS X does have the `daemon` function https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/daemon.3.html

...but it's deprecated. Here's me trying to compile a C program that uses it:

```
$ clang -Wall use_daemon.c
use_daemon.c:4:5: warning: 'daemon' is deprecated: first deprecated in macOS 10.5 - Use posix_spawn APIs instead. [-Wdeprecated-declarations]
    daemon(0, 0);
    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/stdlib.h:294:6: note: 'daemon' has been explicitly marked deprecated here
int      daemon(int, int) __DARWIN_1050(daemon) __OSX_AVAILABLE_BUT_DEPRECATED_MSG(__MAC_10_0, __MAC_10_5, __IPHONE_2_0, __IPHONE_2_0, "Use posix_spawn APIs instead.") __WATCHOS_PROHIBITED __TVOS_PROHIBITED;
         ^
1 warning generated.
```

(macOS 10.5 was [released in 2007](https://en.wikipedia.org/wiki/Mac_OS_X_Leopard#:~:text=leopard%20was%20released%20on%20october%2026%2C%202007) 🤯)

AS a result, it's [hidden in the `libc` crate](https://github.com/rust-lang/libc/blob/96c85c1b913604fb5b1eb8822e344b7c08bcd6b9/src/unix/bsd/apple/mod.rs#L5064-L5067): 

```rust
    #[doc(hidden)]
    #[deprecated(since = "0.2.49", note = "Deprecated in MacOSX 10.5")]
    #[cfg_attr(not(target_arch = "aarch64"), link_name = "daemon$1050")]
    pub fn daemon(nochdir: ::c_int, noclose: ::c_int) -> ::c_int;
```

> The right fix is to file a PR against nix adding `darwin` to the list of OSs that support `daemon`. I'd encourage you to give that a go, as it should be pretty obvious if it compiles and links on OS X, and then future versions of nix (and pizauth) will "just work (TM)".

Trying to do this will cause `nix` to emit the `deprecated()` warning above:

```
error: use of deprecated function `libc::daemon`: Deprecated in MacOSX 10.5
   --> src/unistd.rs:970:30
    |
970 |     let res = unsafe { libc::daemon(nochdir as c_int, noclose as c_int) };
    |                              ^^^^^^
    |
note: the lint level is defined here
   --> src/lib.rs:46:24
    |
46  | #![cfg_attr(test, deny(warnings))]
    |                        ^^^^^^^^
    = note: `#[deny(deprecated)]` implied by `#[deny(warnings)]`

error: could not compile `nix` due to previous error
```

macOS wants you to use [`posix_spawn(3)`](https://man7.org/linux/man-pages/man3/posix_spawn.3.html) instead of `daemon(3)`, but I believe support for it in `nix` is an open issue: https://github.com/nix-rust/nix/issues/1740

---

So no quick and easy fix for this, unfortunately :/

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -38,6 +38,9 @@ urlencoding = "2"
 pledge = "0.4"
 unveil = "0.3"
 
+[target.'cfg(target_os="macos")'.dependencies]
+libc = "0.2"
+
 [profile.release]
 opt-level = 3
 debug = false

--- a/src/compat/daemon.rs
+++ b/src/compat/daemon.rs
@@ -1,0 +1,29 @@
+//! Provides daemon(3) on macOS.
+
+// We provide our own wrapper for daemon on macOS because nix does not export one for macOS.  This
+// is *probably* why nix does not support daemon(3) on macOS:
+//
+//  - nix will not compile on macOS, due to errors
+//  - ... nix compiles with #[deny(warnings)], which treats warnings as errors
+//  - libc emits a deprecation warning for daemon(3) on macOS [1]
+//  - ... because daemon(3) has been deprecated in macOS since Mac OS X 10.5
+//  - ... presumably because Apple wants you to use launchd(8) instead [2].
+//  - Therefore, this deprecation warning is treated as an error in nix
+//
+// [1]: https://github.com/rust-lang/libc/blob/96c85c1b913604fb5b1eb8822e344b7c08bcd6b9/src/unix/bsd/apple/mod.rs#L5064-L5067
+// [2]: https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html
+//
+// This module essentially reimplements nix's daemon wrapper on macOS, but allows deprecation
+// warnings.
+//
+// See: https://github.com/ltratt/pizauth/issues/3
+use libc::c_int;
+#[allow(deprecated)]
+use libc::daemon as libc_daemon;
+use nix::errno::Errno;
+
+pub fn daemon(nochdir: bool, noclose: bool) -> nix::Result<()> {
+    #[allow(deprecated)]
+    let res = unsafe { libc_daemon(nochdir as c_int, noclose as c_int) };
+    Errno::result(res).map(drop)
+}

--- a/src/compat/mod.rs
+++ b/src/compat/mod.rs
@@ -1,0 +1,11 @@
+//! Shims to provide compatibility with different systems.
+
+// nix does not support daemon(3) on macOS, so we have to provide our own implementation:
+#[cfg(target_os = "macos")]
+mod daemon;
+#[cfg(target_os = "macos")]
+pub use daemon::daemon;
+
+// Use nix's daemon(3) wrapper on other platforms:
+#[cfg(not(target_os = "macos"))]
+pub use nix::unistd::daemon;

--- a/src/main.rs
+++ b/src/main.rs
@@ -1,5 +1,6 @@
 #![allow(clippy::derive_partial_eq_without_eq)]
 
+mod compat;
 mod config;
 mod config_ast;
 mod server;
@@ -15,11 +16,11 @@ use std::{
 
 use getopts::Options;
 use log::error;
-use nix::unistd::daemon;
 #[cfg(target_os = "openbsd")]
 use pledge::pledge;
 use server::sock_path;
 
+use compat::daemon;
 use config::Config;
 use user_sender::show_token;
 

